File: packages/unity/com.ishant89.geonavkit/Runtime/TrailSampler.cs
//
//  TrailSampler.cs
//  GeoNavKit
//
//  Pure geometry for placing history-trail dots along a recent path. Returns
//  coordinates only — no rendering dependency.
//

using System;
using System.Collections.Generic;

namespace GeoNavKit
{
    public static class TrailSampler
    {
        private const double MetersPerNM = Distance.MetersPerNauticalMile;

        /// <summary>
        /// <paramref name="count"/> positions evenly spread along the recent path
        /// (last 8 samples). Ordered oldest → newest.
        /// </summary>
        public static List<GeoCoordinate> EqualSpaced(IReadOnlyList<GeoCoordinate> history, int count)
        {
            if (history == null) throw new ArgumentNullException(nameof(history));

            var window = Suffix(history, 8);
            if (window.Count < 2) return Suffix(window, count);

            var cum = new double[window.Count];
            for (int i = 1; i < window.Count; i++)
            {
                cum[i] = cum[i - 1] + Geo.DistanceMeters(window[i - 1], window[i]);
            }
            double total = cum[cum.Length - 1];
            if (total <= 0) return new List<GeoCoordinate> { window[window.Count - 1] };

            var result = new List<GeoCoordinate>(count);
            for (int k = 1; k <= count; k++)
            {
                double target = total * k / count;
                int seg = window.Count - 2;
                for (int i = 0; i < window.Count - 1; i++)
                {
                    if (cum[i + 1] >= target) { seg = i; break; }
                }
                int seg1 = Math.Min(seg + 1, window.Count - 1);
                double segLen = cum[seg1] - cum[seg];
                double t = segLen > 0 ? (target - cum[seg]) / segLen : 0.0;
                result.Add(new GeoCoordinate(
                    window[seg].Latitude + t * (window[seg1].Latitude - window[seg].Latitude),
                    window[seg].Longitude + t * (window[seg1].Longitude - window[seg].Longitude)));
            }
            return result;
        }

        /// <summary>
        /// Exactly <paramref name="count"/> positions spaced <paramref name="spacingNM"/> NM
        /// apart, walking backward from the newest point; short history is projected
        /// backward. Oldest → newest.
        /// </summary>
        public static List<GeoCoordinate> FixedSpaced(IReadOnlyList<GeoCoordinate> history,
                                                      int count, double spacingNM)
        {
            if (history == null) throw new ArgumentNullException(nameof(history));
            if (history.Count < 2) return new List<GeoCoordinate>();

            double spacingMeters = spacingNM * MetersPerNM;

            var result = new List<GeoCoordinate>(count);
            double walked = 0.0;
            int dotNum = 1;
            int i = history.Count - 1;

            while (i > 0 && result.Count < count)
            {
                var segTo = history[i];
                var segFrom = history[i - 1];
                double segLen = Geo.DistanceMeters(segFrom, segTo);

                while (dotNum * spacingMeters <= walked + segLen && result.Count < count)
                {
                    double t = segLen > 0 ? (dotNum * spacingMeters - walked) / segLen : 0.0;
                    result.Add(new GeoCoordinate(
                        segTo.Latitude + t * (segFrom.Latitude - segTo.Latitude),
                        segTo.Longitude + t * (segFrom.Longitude - segTo.Longitude)));
                    dotNum += 1;
                }

                walked += segLen;
                i -= 1;
            }

            if (result.Count < count)
            {
                double backBearing = Geo.Bearing(history[1], history[0]);
                while (result.Count < count)
                {
                    double extra = dotNum * spacingMeters - walked;
                    result.Add(Geo.Offset(history[0], extra, backBearing));
                    dotNum += 1;
                }
            }

            result.Reverse();
            return result;
        }

        /// <summary>The last <paramref name="n"/> elements, or all of them if there are fewer.</summary>
        private static List<GeoCoordinate> Suffix(IReadOnlyList<GeoCoordinate> source, int n)
        {
            int start = Math.Max(0, source.Count - n);
            var slice = new List<GeoCoordinate>(source.Count - start);
            for (int i = start; i < source.Count; i++) slice.Add(source[i]);
            return slice;
        }
    }
}

File: packages/react-native-geonavkit/src/trailSampler.ts
/**
 * Pure geometry for placing history-trail dots along a recent path. Returns
 * coordinates only — no rendering dependency.
 */

import { bearing, distanceMeters, offset } from './geo.js';
import type { Coordinate } from './types.js';

const M_PER_NM = 1852;

/**
 * `count` positions evenly spread along the recent path (last 8 samples).
 * Ordered oldest → newest.
 */
export function equalSpaced(
  history: readonly Coordinate[],
  count: number
): Coordinate[] {
  const window = history.slice(-8);
  if (window.length < 2) return window.slice(-count);

  const cum = new Array<number>(window.length).fill(0);
  for (let i = 1; i < window.length; i++) {
    cum[i] = cum[i - 1]! + distanceMeters(window[i - 1]!, window[i]!);
  }
  const total = cum[cum.length - 1]!;
  if (total <= 0) return [window[window.length - 1]!];

  const result: Coordinate[] = [];
  for (let k = 1; k <= count; k++) {
    const target = (total * k) / count;
    let seg = window.length - 2;
    for (let i = 0; i < window.length - 1; i++) {
      if (cum[i + 1]! >= target) {
        seg = i;
        break;
      }
    }
    const seg1 = Math.min(seg + 1, window.length - 1);
    const segLen = cum[seg1]! - cum[seg]!;
    const t = segLen > 0 ? (target - cum[seg]!) / segLen : 0;
    result.push({
      latitude:
        window[seg]!.latitude +
        t * (window[seg1]!.latitude - window[seg]!.latitude),
      longitude:
        window[seg]!.longitude +
        t * (window[seg1]!.longitude - window[seg]!.longitude),
    });
  }
  return result;
}

/**
 * Exactly `count` positions spaced `spacingNM` NM apart, walking backward from
 * the newest point; short history is projected backward. Oldest → newest.
 */
export function fixedSpaced(
  history: readonly Coordinate[],
  count: number,
  spacingNM: number
): Coordinate[] {
  if (history.length < 2) return [];
  const spacingMeters = spacingNM * M_PER_NM;

  const result: Coordinate[] = [];
  let walked = 0;
  let dotNum = 1;
  let i = history.length - 1;

  while (i > 0 && result.length < count) {
    const segTo = history[i]!;
    const segFrom = history[i - 1]!;
    const segLen = distanceMeters(segFrom, segTo);

    while (dotNum * spacingMeters <= walked + segLen && result.length < count) {
      const t = segLen > 0 ? (dotNum * spacingMeters - walked) / segLen : 0;
      result.push({
        latitude: segTo.latitude + t * (segFrom.latitude - segTo.latitude),
        longitude: segTo.longitude + t * (segFrom.longitude - segTo.longitude),
      });
      dotNum += 1;
    }

    walked += segLen;
    i -= 1;
  }

  if (result.length < count) {
    const backBearing = bearing(history[1]!, history[0]!);
    while (result.length < count) {
      const extra = dotNum * spacingMeters - walked;
      result.push(offset(history[0]!, extra, backBearing));
      dotNum += 1;
    }
  }

  return result.reverse();
}

/** Namespace object mirroring the Swift `TrailSampler` enum. */
export const TrailSampler = { equalSpaced, fixedSpaced } as const;
